### ISSUE TITLE
Support [Parallelizable] across processes rather than threads

### ISSUE THREAD

**Issue by roji** [nunit/nunit#1154]
Playing around with NUnit 3 and reading the [parallelization docs](https://github.com/nunit/nunit/wiki/Parallel-Test-Execution), it seems that the new in-assembly parallelization (with [Parallelizable]) is only possible with threads and not processes, whereas the (older-style?) multiple assembly parallelization uses processes.

I was wondering if there was a particular reason for this. Parallelizing tests across processes rather than threads allows people to immediately run most tests without working about isolation/interference via statics - you only need to worry about external resources (e.g. filesystem, databases). Processes also seem to add very minimal overhead when compared to threads, so the decision to restrict the new [Parallelizable] system to threads only seems strange. @CharliePoole this may be what you referred to in [this comment](https://github.com/nunit/nunit/issues/164#issuecomment-94315627), but I didn't see a separate issue so I'm opening one.

In general, having two different parallelization "features" within NUnit seems like a confusing state of affairs, I hope you guys have some plan to converge these at some point.

Regardless (and right away), documentation on the two forms of parallelization is somewhat confusing... On the [console runner command-line docs](https://github.com/nunit/nunit/wiki/Console-Command-Line), there's nothing that specifies that `--process`, `--inprocess` and `--agents` are for the older-style parallelization (assembly-based), whereas the --workers options is specific to the new [Parallelizable] feature. You may consider clearing that up.

**Comment by rprouse** [comment]
@roji parallelizing tests across processes is an interesting idea. We would need to consider the consequences of creating that many processes for large test suites. Some users already run into memory constraints running many test assemblies in multiple processes. I think it would have to be something that users opt into.

I should also clear up that running assemblies in parallel across processes is actually a new feature in NUnit 3.0. NUnit 2.6 was capable of running each assembly in a different process, but it did so sequentially.

**Comment by roji** [comment]
@rprouse thanks for the response. I agree using processes is definitely something to opt into.

Just to make the motivation here clear: I have several projects where making the test suites threadsafe is either impossible or a very serious effort; in some cases the project I'm testing itself isn't threadsafe (uses statics), in others the test suite itself isn't threadsafe. While memory is definitely a consideration, being able to just turn on parallelization on any testsuite without worrying about statics would probably help people out a lot.

Sorry for the misunderstanding about parallel assemblies, I was under the impression this was supported before. In any case, it's quite confusing to have 2 different parallelization features with varying command-line switches and even behavior (threads vs. processes).

**Comment by rprouse** [comment]
@roji, I have actually been asked a few times at work about the difference between the agents and workers command lines. I have updated the documentation to hopefully make it a bit clearer.

I understand your motivation for multi-process parallization. I see the same thing a lot and think it is a good idea. I am not assigning this to a milestone though because I would like to see more people contribute to the idea and because running test methods in parallel is a higher priority for us. Currently we only support running in parallel down to the test fixture.

**Comment by roji** [comment]
Sure thing @rprouse, test methods in parallel is definitely very important. And thanks for the doc update, it's definitely much better.

**Comment by CharliePoole** [comment]
@rprouse It sounds to me as if @roji is asking for a **third** sort of parallelization - or at least a new subtype of framework parallelization.

@roji If you think of it as process versus thread parallelism, it may be a bit confusing, and maybe our docs are helping lead to that conclusion. I think of it as

1. Assembly level parallelization - the simplest thing to do - except for any common external resources, there is no conflict and no need for coordination across the processes.

2. Test or framework level parallelization - the complexity for the user depends on each test, so we use per-test attributes. If I understand you correctly, you would like some of the tests to run in a separate process. This might use something like [Parallelizable, SeparateProcess] on the class or the method. Have I understood?

Assuming I got it right...

It seems to me that the running of a test in a separate process is orthogonal to the question of parallelism. That is, if you could mark a test as requiring a separate process, the feature should simply work either single threaded or multi-threaded. Naturally, we would have to ensure that any dependencies between tests continued to work even if some were run out of process. The easiest way to do this may be to continue to use a local thread to monitor the process.

I'm pretty sure we have discussed this before but I don't find an issue anywhere.

The same thing could also be done using separate AppDomains. Perhaps we are talking about an IsolationAttribute.

**Comment by roji** [comment]
@CharliePoole and @rprouse, here's how I see it (am undoubtedly missing some points)...

First, I'm not seeing any reason why assembly-level parallelization and test parallelization (with [Parallelizable]) should be two fundamentally distinct things: at the end of the day assemblies are containers of fixtures which themselves are containers of tests (which themselves are containers of test cases). It seems somewhat arbitrary to say that one container level (i.e. the assembly) gets parallelized via processes, while others (i.e. the fixture, the test) get parallelized via threads.

So in a sense I agree with @CharliePoole in that the process vs. thread question (how you isolate tests) is orthogonal to the question of parallelism, i.e. determining which tests can run in parallel with which others. In an ideal world we'd have ways of expressing what can run concurrently with what, at all levels (assembly, fixture, test), and and also ways of expressing what isolation mechanism you use. 

So in my mind it's not three sorts of parallelization, more like one sort of parallelization with two orthogonal "axes" or concerns.

BTW it's also true that AppDomain could represent another isolation level between threads and processes, although it probably doesn't differ from processes much; note also that AppDomains aren't supported in CoreCLR. So I'd say the AppDomains values may be limited, and they definitely shouldn't be the *only* alternative to threads.

**Comment by CharliePoole** [comment]
@roji I agree with you that the two kinds of parallelism are not fundamentally different at theoretical level, but they have important practical differences:

* If you are running test assemblies for which you don't have the source, you can't use attribute-based parallelism. This is a corner case, of course, but it applies to situations where test assemblies are distributed with a product.

* Parallelism between assemblies is implemented in the engine layer of NUnit, while parallelism within an assembly is the job of the framework. That's not to say we can't launch processes from the framework as well, of course.

* The engine has the job of deciding what framework is used by a particular test assembly. It does that by choosing among installed drivers. If the chosen framework is NUnit 3.0, then framework parallelism may be used. Assembly-level parallelism is the only way that we can run tests in parallel if they are written against frameworks that **don't** support parallelism. Currently, you can do that for NUnit V2 tests run under the 3.0 console. In future, folks may write drivers for foreign frameworks to take advantage of this feature.

**Comment by roji** [comment]
@CharliePoole I wasn't at all aware of the engine/framework design and division of labor, I also definitely appreciate the need to support NUnit 2 and other possible frameworks.

Going back to the original request, I guess it may simply be a matter of introducing the concept of isolation level (i.e. threads vs. processes) to the framework-level parallelism, allowing tests within the same assembly to execute in different processes. You would then have two mechanisms supporting process isolation - at the engine level for different assemblies and at the framework for in-assembly (explaining that in the docs might not be trivial :)). I guess that's what you were saying in your second point.

**Comment by timotei** [comment]
+1 for allowing Parallelizable to work at process-level not only at thread level.

We are also having a lot of tests that can't properly run parallelized at thread level (or the effort is very big). Up until now, we've solved the issue by executing *multiple*, separate runs of tests in specific namespaces, in parallel (e.g., via gallio). The problem is that we need to merge the results and so on.

It would be very nice to have this built-in in the NUnit framework. If the idea is approved we could help with contributions via PRs if needed.

**Comment by rprouse** [comment]
Thanks for the idea, but this isn't in keeping with our vision of where we want to take NUnit, so I am closing.

**Comment by wrummler** [comment]
@rprouse It's been a while since this was closed, so I figured it couldn't hurt asking: Does the current roadmap of NUnit still preclude this feature?

I can describe my small team's scenario and why we would benefit from being able to declare that certain tests should each run in their own process.

We have some tests (about 500) that depend on the process-wide current directory. (For the record, we know this dependency is not a good thing, especially in tests. We are in the middle of incrementally transitioning a modestly large (250 kloc), _very_ old (10--40 years), and previously test-less codebase to better architecture and tooling.)

Each of these ~500 tests takes 1--10 seconds to run on average, with just a few being quite a bit longer. (Again for the record, these are regression/integration tests that act on a couple dozen highly heterogeneous subsystems of the overall application, and they are unfortunately relatively long-running, for now.)

Currently, due to the fact that each test must be executing with a distinct current directory during the entire test execution, parallelizing with the `Parallelizable` attribute leads to test failures.

Other than the process-wide current directory, there are no other shared dependencies that cause these failures. We know this because we typically run the tests successfully in parallel with an in-memory implementation of our file-system abstraction.

The reason we sometimes want to run the tests in parallel with the on-disk shim implementation of our file-system abstraction is that this reflects our application's "legacy mode", which is currently useful for our debugging efforts in certain situations.

All this being said, we do not find the current lack of this feature to be anything remotely close to a deal-breaker, but we did want to add our support to its eventual appearance in NUnit.

Thanks.
